fsspec/implementations/cache_metadata.py

```diff
@@ -0,0 +1,210 @@
+    ) -> Literal[False] | tuple[Detail, str]:
+        """If path is in cache return its details, otherwise return ``False``.
+
+        Takes an optional callback so that the caller can choose to ignore
+        possible matches if, for example, they are too old.
+        """
+        for (fn, base, _), cache in zip(self._scan_locations(), self.cached_files):
+            if path not in cache:
+                continue
+            detail = cache[path].copy()
+            if ignore is not None and ignore(detail):
+                continue
+            fn = os.path.join(base, detail["fn"])
+            if os.path.exists(fn):
+                return detail, fn
+        return False
+
+    def clear_expired(self, expiry_time: int) -> Iterator[str]:
+        """Remove expired metadata from the cache.
+
+        Yields names of files corresponding to expired metadata. Caller is
+        responsible for deleting these files.
+        """
+        for path, detail in self.cached_files[-1].copy().items():
+            if time.time() - detail["time"] > expiry_time:
+                fn = detail.get("fn", "")
+                if not fn:
+                    raise RuntimeError(
+                        f"Cache metadata does not contain 'fn' for {path}"
+                    )
+                fn = os.path.join(self._storage[-1], fn)
+                yield fn
+                self.cached_files[-1].pop(path)
+
+        if self.cached_files[-1]:
+            cache_path = os.path.join(self._storage[-1], "cache")
+            with atomic_write(cache_path) as fc:
+                pickle.dump(self.cached_files[-1], fc)
+
+    def close_file(self, f: Any, path: str) -> None:
+        """Perform side-effect actions when closing a cached file.
+
+        The actual closing of the file is the responsibility of the caller.
+        """
+        # File must be writeble, so in self.cached_files[-1]
+        c = self.cached_files[-1][path]
+        if c["blocks"] is not True and len(c["blocks"]) * f.blocksize >= f.size:
+            c["blocks"] = True
+
+    def empty(self) -> bool:
+        """Return ``True`` if metadata of the writable storage is empty"""
+        return not self.cached_files[-1]
+
+    def load(self) -> None:
+        """Load all metadata from disk and store in ``self.cached_files``"""
+        cached_files = []
+        for fn, _, _ in self._scan_locations():
+            if os.path.exists(fn):
+                with open(fn, "rb") as f:
+                    # TODO: consolidate blocks here
+                    loaded_cached_files = pickle.load(f)
+                    for c in loaded_cached_files.values():
+                        if isinstance(c["blocks"], list):
+                            c["blocks"] = set(c["blocks"])
+                    cached_files.append(loaded_cached_files)
+            else:
+                cached_files.append({})
+        self.cached_files = cached_files or [{}]
+
+    def pop_file(self, path: str) -> None:
```

Comment:
Docstring: remove local copy of file and its metadata

Reply:
Done. Also moved the deletion of the cached file up to `CachedFileSystem` as that is consistent with the class responsibilities.